### PR TITLE
Fixed Idempotency problem with struct end

### LINKED ISSUE #2592: Idempotency problem with `struct end`
Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAXAngB0WAGmALxgDO6ATgK7TpiIB28AOg61rmAJrFgCWDdIgoAzAIbQ8jeJAA04KKUTp0AgOalUYANoKIoCIagASDsjSQB6OXsORjANzEAbKuZ22jABnlGjkAEkmRnQAZT4AL2RfPwgAFk8wAF1bAF8Y-UT7My1LQRtY%2BydXd11YiABGDL9IAFkxAA8AGQFEJsY1dAALAvKKgCYvRJSjdNsDGtMcd0hpAHkRFoZEUKxnaKzHFzctMvL%2B6v8AUSYFpY3yqGcRSGG06on-KdxcgCNYWGdep%2BKdtD3YgBmQ52IJKCjoABiAhcADlEAB3Zytb5%2BShuO6jB6bHIWd6fVEQIrbUqJeIgomhbASRAAIUQIlgFEQAAUxBQxABbZTCQmGdGITGGMZGR52Z4zfFfEHEkq7MlgACsFKgVJp9MZzKasARwmgYiUQQcsH1qlgDD5EAFQogIsMYqJEreH2lm1%2BpMuADYVZA1ZINUzEABVbC4Cj6w0MY2mvjmy1gcTOJQ2sB2zKFJ14l182V-DyXADsPr9dIZgYAws4DaRy%2BbyNRaEz44nk7ERsLsRncVApTmtnL-gqABzF6n%2BsvM2qITmvXkqluCtv3cY46bOgky-t5gF%2BACco-VE8QteccfnLlbfnbts7k27kF7m-d8suFR8Ct9Y8QAEEREIKLWnKcmI8bWkuWIrl2a5Zhuboki%2BfRVB%2BJYBsyoTTnwcCnha55JouV7LqKq4vDBroZs%2Bg6vgcyFfr%2B-7oZymGfGeCpgQREFEVBJE9tmT7wZRfTAjRNLfhQsBUEwAAiiDIox-6gdQ%2BFGNeqa3k895WH2FH5n0cQ%2BvUDQBCIAAqXSMKEXRMugADqfDwN08ZDOBHaQXe0FQJpfEDjpsQVMqH4GUZpmMEceEWVZtn2T0Kqek57EuZxbncXk1heduCoVN6AWNEEIh8A0cxhmI6BMkcDTYMypCkLGOHDnFymEfaxEzJ5cHeTuRgVEW2UNAASogcAUPAkUOSqcT1YYKlpmADomBp%2BRpR6fQjj1-WDfAsJUDOwgLAEQicpoKoVCm02zdk7mQJyVDOKoyLLJCTLASowjGWui0Ib5%2B4fmtTLwLU123a0D0UE98k%2BtAXTshI-4APoInZDknWp4rzalbXpZcgz6Y0okcpgcwUC05AjdFdVI656kXa15H8T5fj9Ehlx1DjFB4wTRPoJt20ULt%2B2HRl5OJZTyVXTdfB3YgwOgy9b3o0tsT9NRTO42I%2BOE3w5D-WLEtS8VYMfhDUO0MIcMIz0gtNVxLULXLH300JTMGQAaiStICPA6gk-GQ4TRAU3I46qNabTHWGP0ek9ZCEm0DVbtMJ7ZvxuNFvpkl1tozT7UKv0-mO40kmwOgADiyhlRViBVTVXsqj7Kczc165kZM2mhxA-RZY7APi60tKntAADWtIcgPyikHMDChFyx6fFttWXAuddnZmPGwZnGP7N1TPwkiPfKLqjCvbg0l5QwfBmgw35MFO3P85cbENRxltp43wdZ5jK3K8iagMFHDAx%2Ba4Rv7FSoMyYysAghIEEMVGqzYLxKUmo1VOwtJS8VtgJBWX1P7-gYNAhwbRzRqGvrOCgx8BBnxqrfcoC9nI3gpijC6j40F0yMICd8ncxbnCaFyV48AxCVlgNVBgaht4S1gXhReDdSKv3XkCRm5RIBl2EHwbkUDnAAGlECIGwBAkIQQh5iD-qTeecCJFWxfu9dBfhARK3kb3Ax-dzikFEuJJg8JyCID%2Bl3CWZdKrVTrApDENDVJ0MDgw1Ba95ZWIdrY9kqFECSQ1tACgyiYRCHgEGU%2B5ppLQCrByc%2BYjLwPwSk-ZB5imGtzAICCOTNFHJJUegFwqwxJUHrFQbAqxMDrAKfAv2iD65mIsNTZuIcFSAlzvIjRWiDJc2IQsOxDAHGtEob5Lwvtkh9KXveRhES7YsI7vIppmF0C1AQBcKhJigmpAUEkGIlhSCQmqloBeqQgA)

#### Code

```fsharp
type X = struct end

type Y = interface end
```
            

#### Result

```fsharp
type X =


    struct

    end

type Y =
    interface
    end

```
            
#### Problem description

Every time you format the code additional new lines get added.

#### Extra information

- [ ] The formatted result breaks my code.
- [ ] The formatted result gives compiler warnings.
- [x] I or my company would be willing to help fix this.

#### Options

Fantomas main branch at 2022-10-19T15:02:12Z - 25cb8d54f9677b4920022b02295e9d00b5eacbaa

Default Fantomas configuration

<sub>Did you know that you can ignore files when formatting from fantomas-tool or the FAKE targets by using a [.fantomasignore file](https://fsprojects.github.io/fantomas/docs/end-users/IgnoreFiles.html)?</sub>

### PR DESCRIPTION
**Pull request by stefan-schweiger**
Fixes #2592

copied the code from the interface/class case

### CODE REVIEW

File: src/Fantomas.Core/CodePrinter.fs [hunk 1]
```diff
@@ -3084,12 +3084,16 @@ and genTypeDefn (TypeDef (ats, px, leadingKeyword, ao, tds, tcs, equalsRange, td
 
         typeName
         +> opt sepNone impCtor genMemberDefn
-        +> genEq SynTypeDefn_Equals equalsRange
+        +> leadingExpressionIsMultiline (opt sepNone impCtor genMemberDefn) (fun isMulti ctx ->
```
Comment (stefan-schweiger) [inline src/Fantomas.Core/CodePrinter.fs:3087]: all tests work, but are there any other implications about `genEq SynTypeDefn_Equals equalsRange` which I don't see?

File: src/Fantomas.Core/CodePrinter.fs [hunk 2]
```diff
@@ -3084,12 +3084,16 @@ and genTypeDefn (TypeDef (ats, px, leadingKeyword, ao, tds, tcs, equalsRange, td
 
         typeName
         +> opt sepNone impCtor genMemberDefn
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:3086]: If there is an `impCtor`, it will be printed twice here.
For example:
```fsharp
type NameStruct() =
    struct
    end
```
becomes
```fsharp
type NameStruct()() =
    struct
    end
```

but there is a bit more to this problem. I'll write it up in the summary.

File: src/Fantomas.Core/CodePrinter.fs (hunk 1 above)
Comment (nojaf) [reply src/Fantomas.Core/CodePrinter.fs:3087]: `genEq` shouldn't really be touched here.
The problem is a bit of a different nature.

**Review by nojaf — CHANGES_REQUESTED** [review]
Hello again Stefan! Very cool to see you dive into the problems you find.
Much appreciated.

For this case, I would like to go back to the drawing board.
There are two problems:
- If there are no members inside the struct there is an additional empty line. (which you solve by `onlyIf (List.isNotEmpty others) sepNln` and is correct!)
- If there is a newline (or trivia in general) between an empty `struct` `end`, the closest anchor is the `=` token (As content after).

We currently don't have the `struct` or `end` keywords as a TriviaNode to anchor.

Hypothetically, if we did, the newline between `struct` and `end` would be anchored to the `end` keyword. There would be an `genTriviaFor SynTypeDefn_ObjectModel_End endRange !- "end"` call and that would solve the larger problem.

The range of the [ObjectModel](https://fsprojects.github.io/fantomas/reference/fsharp-compiler-syntax-syntypedefnrepr.html#ObjectModel) does include both keywords.
We could safely assume when [SynTypeDefnKind](https://fsprojects.github.io/fantomas/reference/fsharp-compiler-syntax-syntypedefnkind.html) is a `Struct` that the start of that range is the start of `struct` and the end, well is the end of `end`.

This means we can create trivia nodes in a similar fashion as @dawedawe is doing in https://github.com/fsprojects/fantomas/pull/2591.

In ASTTransformer, would could extract the information in `visitSynTypeDefnRepr`:
pseudo code:
```fsharp
and visitSynTypeDefnRepr (stdr: SynTypeDefnRepr) : TriviaNode list =
    match stdr with
    | SynTypeDefnRepr.ObjectModel (kind, members, range) ->
        let structNodeAndEndNode =
            match kind, range with
            | SynTypeDefnKind.Struct, StartRange 6 (mStruct, _) & EndRange 3 (mEnd, _) ->
                [ mkNode SynTypeDefn_ObjectModel_Struct mStruct
                  mkNode SynTypeDefn_ObjectModel_End mEnd ]
            | _ -> []
        
        [ yield! Option.toList (visitSynTypeDefnKind kind)
          yield! structNodeAndEndNode
          yield! visitSynMemberDefns members ]
    | SynTypeDefnRepr.Simple (simpleRepr, _) -> visitSynTypeDefnSimpleRepr simpleRepr
    | SynTypeDefnRepr.Exception exceptionRepr -> [ visitSynExceptionDefnRepr exceptionRepr ]
```
* `EndRange` does not exist in `RangeHelpers` but you would get the idea.
Nor do `SynTypeDefn_ObjectModel_Struct` and `SynTypeDefn_ObjectModel_End` in `TriviaTypes`.

In CodePrinter, you need to exact ranges:

```fsharp
| ObjectModel (TCSimple TCStruct as tdk, MemberDefnList (impCtor, others), RangePatterns.StartRange 6 (mStruct, _) & RangePatterns.EndRange 3 (mEnd, _)) ->
```

To then use `mStruct` and `mEnd` within a `genTriviaFor` call.

---

I'm aware this is a lot of ground to cover, but that would be the ideal solution.

We would understandably, also accept the single `onlyIf (List.isNotEmpty others) sepNln` fix. That at least will solve your original problem I think.

File: src/Fantomas.Core.Tests/StructTests.fs [hunk 3]
```diff
@@ -127,3 +127,24 @@ namespace meh
...
+    formatSourceString
+        false
+        """
+type NameStruct =
+
+    struct
+
+    end
+"""
+        { config with MaxValueBindingWidth = 120 }
```
Comment (nojaf) [inline src/Fantomas.Core.Tests/StructTests.fs:142]: For future reference: `MaxValueBindingWidth` has no influence on this test, it is preferred not to be included in the test.

**Comment by nojaf** [comment]
Sure, I see now why you changed
```fsharp
        +> leadingExpressionIsMultiline (opt sepNone impCtor genMemberDefn) (fun isMulti ctx ->
            if isMulti && ctx.Config.AlternativeLongMemberDefinitions then
                sepEqFixed ctx
            else
                sepEq ctx)
```
to begin with.

I don't think we really need the `nonClassRange`, if it is a `TCInterface` it wouldn't
have an `impCtor`.
So, I think you can even further simplify it to

```fsharp
    | ObjectModel (TCSimple (TCInterface | TCClass | TCStruct) as tdk, MemberDefnList (impCtor, others), _) ->
        typeName
        +> leadingExpressionIsMultiline
            (optSingle (fun md -> sepSpaceBeforeClassConstructor +> genMemberDefn md) impCtor)
            (fun isMulti ctx ->
                if isMulti && ctx.Config.AlternativeLongMemberDefinitions then
                    sepEqFixed ctx
                else
                    sepEq ctx)
        +> indent
        +> sepNln
```

There still would be a problem left in that area ([sample](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAXAngB0WAGgBQCUYAvADoBOFAdhBNADYCGAzm9XQ2APS9gAtogAWXeg0S14kADTgobROnQBLWgHM2qMAG0FEUDwiQAJFlw7I69HIMMzANxZMArsjT7jDAAzzvJgCS0lLoAMqqAF7I-gEALPYQALr2AL6xholQ5jgeUDZ23k4u7jpeAQCMGcaQALIsAB4AMuqITVIa6CKFlQBMPlkpPOn2RjU5lmiQUvAA8gBmLbSIYVhMMVnFbnnl3r3VPJAAotILSxsBUEzzkINp1WOHE3mQAEawsEw9T87bZVkQADMBwcwSUVHQADF1C4AHKIADuTFa32M6Co7juwwem2eVnen1RJlMv1KngBYDiIJMYWwLGgiAAQoh5rAqIgAAosKgsYToRBUIkMdGY7xDBgjHiPBx4qYEr4grZkvQUgCs1KgtPpTJZbLasARAug7EQwUcsGNalgtCFEBFiCxEpxRVlUHlQqVOwpADYNZAtQzmaz2QBVbC4KjGpRmi0sK02jXzFxKR0QSUMaXEiwvd2KkklL2XADsfoDOuDiAAwqwOJXrWwRdB0GzbWAk0wU2L7qNcdn8R8FbjSYWAgAOUt0wO69m1RCCV4C1vtzvGcVp53jPtygce-N-cmXACcE%2B1Qb1daY1qXyYdXexPZdW7dO7zw-%2Blwqfgp-sniAAgvM-JUHWgiCCwrb2qmYDppkj65P2hKvgW76VFU35lme7JhHOqhwJeCYUsut6rt2Uq9vB26IUOyEHn0J4MgBQHYYIuGfFeGqQXeToPpuFHPlRLpvrR3gVMC6G-n%2BVCwK40gACKIMiLFARBGLETwa7QRuTxPtYtC2Eh%2B4qh%2BVLfvUDSBPMAAqIhSGEIhsugADqqjwF0rYDFx648dpfG6fp1GGbsxgVOqpmNBZ1lSEcHYrPZELOa53Qat6HkkfeZFwZM%2BR6buQlGZUvpheZtDzKoDSzBGcZskcDTYOyHCqOxFKjql6mkRm5FZX5uU0flIklkVABKiBwFQ8AJW5GpxK1DAaTBYCZtkOkFAZypBTwFTjkNI1svAsKuPOAoLIE-KCNoGoVFB82LWYOmCK4TBqMiyyQmyYEqAKlnwatI4ice37DaN8C1A9T2tK9VDvcpfrQCI3L0kBAD6CIuW5V1aTKy05T9KF7F%2Blx1I0kk8pgsxUC0DYTUlzUzck7WwbxXUrQFa0Ur0aEE2ZxMsKT5OqA2%2B2HVQx2nedFKXZ5mneZjvn3Y9qjPYgENQ5930s79xi9Ps37c7zFPoCD8uK8rcbQ9%2BsPw02ArI6j3To9LWa%2Bczgm9etDC9GJnONAAagWjLqPA6gaFTrYtfbGWMy8zvjHlbsQL0Jlew0kIyU2jW0P70hByHU202Ac0Y47TPY%2BruOa6FSeybA6AAOLKDVdWIA11o5zT4cdZlOYvqXwma4VnOgwrrSMpe0AANaMjy4-KGwsy0GEvJVp8B0EZcRHtwzPldbmPd9ZrA0E-CSLD8ohpSF9uDyaVtCqPGf7SLOQti5cnFpdxEdb13Akx67bNbQTf5kQaFoCnWgadrQRGAXGVw7JLKwGCEgPScZ07XhihvBanUv6DhdoFNm-0AGPQFLQZBjh9SaEfguKgV91C33Ts-AI69JbXUwQhbBP9cGXEBPjAIdRB7nCaLyV48AWDVlgGwIOR9FaoJXG1dKHdI6sJ6hwgIgIOY8IbgKVQwgkFMAANKIEQNgBBoRgiTxYGA6ma8bzoJuq6N43ccGs04drAmI9zFj3OGwSS0lpDwgbIgYGg9FYN3quI%2BsKlRRvy8h-GW28HHsKcSoz2PDGTckwogWS-NoBUC0TCfk8AQw32tPJZg8N4zSLUrNemGDO6KJxr3HggJE7qNqpo7R6AXCrCkq4BsGJsCrEwOsCpNiWFTGjj8X%2BnCK48P0YYsygtKELDcbQDxrR6EiR8HnAuDslq%2BR3o4jWjT%2B48K6bhA2CALgMOsUwhQSRYjWDYJCcROh16pCAA)) but that is something for another day...

**Comment by stefan-schweiger** [comment]
Completely removing the class check broke some tests, but I think the code now should be fine, please take another look ;)

**Comment by nojaf** [comment]
> Completely removing the class check broke some tests

Yes, that problem was addressed in my previous snippet. The `(optSingle (fun md -> sepSpaceBeforeClassConstructor +> genMemberDefn md) impCtor)` should take care of that.
Please update accordingly.

**Comment by stefan-schweiger** [comment]
Sorry I think I only replaced the first instruction which lead to additional new lines. Code should now work

File: src/Fantomas.Core.Tests/StructTests.fs [hunk 4]
```diff
@@ -127,3 +127,26 @@ namespace meh
 
 type TupleStruct = (struct (string * string))
 """
+
+[<Test>]
+let ``struct empty type`` () =
+    formatSourceString
+        false
+        """
+type NameStruct() =
```
[the remark this answers is not in the document]
Comment (stefan-schweiger) [reply src/Fantomas.Core.Tests/StructTests.fs:136]: Is it ok to add additional tests? Because there could be changes in the future which break the empty line behavior again which wouldn't be caught by current tests

[the remark this answers is not in the document]
Comment (nojaf) [reply src/Fantomas.Core.Tests/StructTests.fs:136]: I wouldn't mind, but add a comment that the expected result is not entirely correct.

If the input is:

```fsharp
type X =
  
    interface
    // comment
    end
```

that exact code should be produced. Which will not be the case with the current changes.
If somebody later ever breaks that test, by having a better result, then that person will at least know what is up.

[the remark this answers is not in the document]
Comment (stefan-schweiger) [reply src/Fantomas.Core.Tests/StructTests.fs:153]: I don't want to go as far as adding tests for broken functionality. I just wanted to add tests for edge cases identified during this PR. See my latest changes :)
